File: libraries/cary/ary1Ccpy.c
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "ary_err.h"
#include "dat_par.h"
#include <ctype.h>

void ary1Ccpy( const char *cin, size_t len, char *cout, int *status ) {
/*
*+
*  Name:
*     ary1Ccpy

*  Purpose:
*     Copy a character string, checking for truncation.

*  Synopsis:
*     void ary1Ccpy( const char *cin, size_t len, char *cout, int *status )

*  Description:
*     The routine copies a character string from one variable to
*     another and checks for truncation of trailing non-blank characters.
*     If such truncation occurs, then an error is reported and a "status"
*     value set.

*  Parameters:
*     cin
*        The input character string.
*     len
*        The length of the output array.
*     cout
*        The array to receive the copied string.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local Variables: */
   char *p1;
   const char *p2;
   int nc;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Copy the string, stopping when the end of the input string is reached
   or the output buffer is filled (leaving room for a trailing null). */
   p1 = cout;
   p2 = cin;
   nc = 0;
   while( nc++ < len - 1 && *p2 ) {
      *(p1++) = *(p2++);
   }

/* Terminate the output string. */
   *p1 = 0;

/* If characters remain uncopied, then test to see if they are blank.
   Report an error if they are not. */
   while( *p2 ){
      if( !isspace( *(p2++) ) ) {
         *status = ARY__TRUNC;
         msgSetc( "STRING", cout );
         errRep( "ARY1_CCPY_STR", "Character string truncated: '^STRING'.",
                 status );
         errRep( " ", "Output character variable is too short to accommodate the"
                 "returned result (possible programming error).", status );
      }
   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Ccpy", status );

}

File: libraries/fary/ary1_ccpy.f
      SUBROUTINE ARY1_CCPY( CIN, COUT, STATUS )
*+
*  Name:
*     ARY1_CCPY

*  Purpose:
*     Copy a character string, checking for truncation.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_CCPY( CIN, COUT, STATUS )

*  Description:
*     The routine copies a character string from one variable to
*     another (following the rules of Fortran character assignment) and
*     checking for truncation of trailing non-blank characters. If such
*     truncation occurs, then an error is reported and a STATUS value
*     set.

*  Arguments:
*     CIN = CHARACTER * ( * ) (Given)
*        The input character string.
*     COUT = CHARACTER * ( * ) (Returned)
*        The output character variable to receive the string.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Copy the string with an assignment statement.
*     -  If there are characters remaining, check they are all blank.
*     If not, then report an error.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     7-JUN-1989  (RFWS):
*        Original version.
*     8-AUG-1989 (RFWS):
*        Split error report into two, to prevent long strings from
*        causing the trailing part of the message to get lost.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Arguments Given:
      CHARACTER * ( * ) CIN

*  Arguments Returned:
      CHARACTER * ( * ) COUT

*  Status:
      INTEGER STATUS             ! Global status

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Copy the string.
      COUT = CIN

*  If characters remain uncopied, then test to see if they are blank.
*  Report an error if they are not.
      IF ( LEN( CIN ) .GT. LEN( COUT ) ) THEN
         IF ( CIN( LEN( COUT ) + 1 : ) .NE. ' ' ) THEN
            STATUS = ARY__TRUNC
            CALL MSG_SETC( 'STRING', COUT )
            CALL ERR_REP( 'ARY1_CCPY_STR',
     :      'Character string truncated: ''^STRING''.', STATUS )
            CALL ERR_REP( 'ARY1_CCPY_TRNC',
     :      'Output character variable is too short to accommodate ' //
     :      'the returned result (possible programming error).',
     :      STATUS )
         END IF
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_CCPY', STATUS )

      END
